File: src/lib/manga-detector.ts
// src/lib/manga-detector.ts
import fs from 'fs-extra';
import path from 'path';
import AdmZip from 'adm-zip';
import { XMLParser } from 'fast-xml-parser';
import { Logger } from './logger';
import { getErrorMessage } from './utils/error';
import { prisma } from './db';

// Default Internal Publisher Dictionary Fallbacks
const DEFAULT_MANGA_PUBLISHERS = [
    "viz media", "kodansha", "yen press", "seven seas", "shueisha", 
    "shogakukan", "tokyopop", "dark horse manga", "vertical", 
    "ghost ship", "denpa", "fakku", "j-novel club", "sublime", 
    "kuma", "ize press", "square enix", "hakusensha", "lezhin"
];

const DEFAULT_WESTERN_PUBLISHERS = [
    "marvel", "dc comics", "image comics", "idw publishing", 
    "dynamite", "boom! studios", "valiant", "archie", 
    "oni press", "titan comics", "vault comics", "awa studios", "humanoids", "2000 ad", "zenescope"
];

const MANGA_CONCEPTS = [
    "manga", "shonen", "seinen", "shojo", "josei", 
    "manhwa", "manhua", "webtoon", "tankobon", "doujinshi"
];

// --- NEW: In-Memory Cache to prevent N+1 DB queries during mass library scans ---
let cachedSettings: { manga: string[], western: string[] } | null = null;
let cacheTimestamp = 0;

async function getDetectorSettings() {
    // Return cache if it is less than 5 minutes old
    if (cachedSettings && Date.now() - cacheTimestamp < 5 * 60 * 1000) {
        return cachedSettings;
    }

    const settings = await prisma.systemSetting.findMany({
        where: { key: { in: ['manga_publishers', 'western_publishers'] } }
    });
    const config = Object.fromEntries(settings.map(s => [s.key, s.value]));
    
    cachedSettings = {
        manga: config.manga_publishers 
            ? config.manga_publishers.split(',').map((s: string) => s.trim().toLowerCase()).filter(Boolean)
            : DEFAULT_MANGA_PUBLISHERS,
        western: config.western_publishers
            ? config.western_publishers.split(',').map((s: string) => s.trim().toLowerCase()).filter(Boolean)
            : DEFAULT_WESTERN_PUBLISHERS
    };
    cacheTimestamp = Date.now();

    return cachedSettings;
}

export async function detectManga(
    comicVineData: any, 
    filePath: string | null = null
): Promise<boolean> {
    
    // Fetch settings (will hit RAM instantly 99% of the time)
    const { manga: mangaPublishers, western: westernPublishers } = await getDetectorSettings();

    // --------------------------------------------------------
    // WATERFALL STEP 1: Check Publisher
    // --------------------------------------------------------
    if (comicVineData?.publisher?.name) {
        const publisher = comicVineData.publisher.name.toLowerCase();
        if (mangaPublishers.some((mp: string) => publisher.includes(mp))) {
            Logger.log(`[Manga Engine] Identified via Publisher: ${publisher}`, 'info');
            return true;
        }
    }

    // --------------------------------------------------------
    // WATERFALL STEP 2: Check ComicVine Concepts
    // --------------------------------------------------------
    if (comicVineData?.concepts && Array.isArray(comicVineData.concepts)) {
        const hasMangaConcept = comicVineData.concepts.some((concept: any) => 
            MANGA_CONCEPTS.includes(concept.name?.toLowerCase())
        );
        if (hasMangaConcept) {
            Logger.log(`[Manga Engine] Identified via ComicVine Concepts`, 'info');
            return true;
        }
    }

    // --------------------------------------------------------
    // WATERFALL STEP 2.5: Western Publisher Hard-Bypass
    // --------------------------------------------------------
    if (comicVineData?.publisher?.name) {
        const publisher = comicVineData.publisher.name.toLowerCase();
        if (westernPublishers.some((wp: string) => publisher.includes(wp))) {
            Logger.log(`[Manga Engine] Bypassing AniList due to Western Publisher: ${publisher}`, 'info');
            
            if (filePath && fs.existsSync(filePath)) {
                const ext = path.extname(filePath).toLowerCase();
                if (ext === '.cbz' || ext === '.zip') {
                    try {
                        const zip = new AdmZip(filePath);
                        const xmlEntry = zip.getEntry("ComicInfo.xml");
                        if (xmlEntry) {
                            const xmlString = xmlEntry.getData().toString("utf8");
                            const parser = new XMLParser();
                            const jsonObj = parser.parse(xmlString);
                            const mangaTag = jsonObj?.ComicInfo?.Manga;
                            if (mangaTag === 'Yes' || mangaTag === 'YesAndRightToLeft') {
                                Logger.log(`[Manga Engine] Override: Identified via ComicInfo.xml`, 'info');
                                return true;
                            }
                        }
                    } catch (e) {}
                }
            }
            return false; 
        }
    }

    // --------------------------------------------------------
    // WATERFALL STEP 3: AniList API Cross-Reference
    // --------------------------------------------------------
    if (comicVineData?.name) {
        try {
            const releaseYear = parseInt(comicVineData.year) || parseInt(comicVineData.start_year) || 0;
            const isAniListMatch = await checkAniList(comicVineData.name, releaseYear);
            
            if (isAniListMatch) {
                Logger.log(`[Manga Engine] Identified via AniList API Match`, 'info');
                return true;
            }
        } catch (e) {
            Logger.log(`[Manga Engine] AniList check failed: ${getErrorMessage(e)}`, 'error');
        }
    }

    // --------------------------------------------------------
    // WATERFALL STEP 4: ComicInfo.xml Check
    // --------------------------------------------------------
    if (filePath && fs.existsSync(filePath)) {
        const ext = path.extname(filePath).toLowerCase();
        if (ext === '.cbz' || ext === '.zip') {
            try {
                const zip = new AdmZip(filePath);
                const xmlEntry = zip.getEntry("ComicInfo.xml");
                if (xmlEntry) {
                    const xmlString = xmlEntry.getData().toString("utf8");
                    const parser = new XMLParser();
                    const jsonObj = parser.parse(xmlString);
                    if (jsonObj?.ComicInfo?.Manga === 'Yes' || jsonObj?.ComicInfo?.Manga === 'YesAndRightToLeft') {
                        return true;
                    }
                }
            } catch (e) {}
        }
    }

    return false;
}

/**
 * Helper Function: Queries AniList GraphQL API with Fuzzy Year Logic
 */
async function checkAniList(title: string, releaseYear: number): Promise<boolean> {
    const query = `
        query ($search: String) {
            Page(page: 1, perPage: 3) {
                media(search: $search, type: MANGA) {
                    title { romaji english }
                    startDate { year }
                    format
                }
            }
        }
    `;

    const response = await fetch('https://graphql.anilist.co', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' },
        body: JSON.stringify({ query, variables: { search: title } })
    });

    if (!response.ok) return false;

    const data = await response.json();
    const mediaResults = data?.data?.Page?.media || [];

    const searchTitle = title.toLowerCase().trim();

    for (const media of mediaResults) {
        const engTitle = media.title?.english?.toLowerCase().trim() || "";
        const romajiTitle = media.title?.romaji?.toLowerCase().trim() || "";

        if (searchTitle === engTitle || searchTitle === romajiTitle) {
            if (releaseYear > 0 && media.startDate?.year) {
                const yearDiff = Math.abs(releaseYear - media.startDate.year);
                if (yearDiff > 4) {
                    Logger.log(`[Manga Engine] AniList match rejected due to Year Mismatch (${releaseYear} vs JP ${media.startDate.year})`, 'info');
                    continue; 
                }
            }
            return true;
        }
    }

    return false;
}

File: omnibus-engine/src/manga_detector.rs
use reqwest::Client;
use sqlx::Row;
use std::time::Duration;

// Built-in fallbacks (parity with manga-detector.ts).
const DEFAULT_MANGA_PUBLISHERS: &[&str] = &[
    "viz media", "kodansha", "yen press", "seven seas", "shueisha", "shogakukan", "tokyopop",
    "dark horse manga", "vertical", "ghost ship", "denpa", "fakku", "j-novel club", "sublime",
    "kuma", "ize press", "square enix", "hakusensha", "lezhin", "kadokawa", "futabasha", "houbunsha",
    "takeshobo", "mag garden", "akita shoten", "shonen gahosha", "nihon bungeisha", "coamix",
    "gee-whiz", "suiseisha", "ascii media works", "ichijinsha", "project-h", "irodori", "eros comix",
];
const DEFAULT_WESTERN_PUBLISHERS: &[&str] = &[
    "marvel", "dc comics", "image comics", "idw publishing", "dynamite", "boom! studios", "valiant",
    "archie", "oni press", "titan comics", "vault comics", "awa studios", "humanoids", "2000 ad", "zenescope",
];

/// Fetches the manga / western publisher lists from settings (or the built-in defaults).
/// Fetch once per scan and pass the lists into `detect_manga` to avoid N+1 queries.
pub async fn get_detector_settings(db: &sqlx::AnyPool) -> (Vec<String>, Vec<String>) {
    let rows = sqlx::query(r#"SELECT key, value FROM "SystemSetting" WHERE key IN ('manga_publishers', 'western_publishers')"#)
        .fetch_all(db)
        .await
        .unwrap_or_default();
    resolve_detector_lists(rows.iter().map(|r| (r.get("key"), r.get("value"))).collect())
}

fn resolve_detector_lists(kv: Vec<(String, String)>) -> (Vec<String>, Vec<String>) {
    let mut manga: Option<Vec<String>> = None;
    let mut western: Option<Vec<String>> = None;
    for (k, v) in kv {
        let list: Vec<String> = v.split(',').map(|s| s.trim().to_lowercase()).filter(|s| !s.is_empty()).collect();
        if !list.is_empty() {
            if k == "manga_publishers" { manga = Some(list); }
            else if k == "western_publishers" { western = Some(list); }
        }
    }

    let manga = manga.unwrap_or_else(|| DEFAULT_MANGA_PUBLISHERS.iter().map(|s| s.to_string()).collect());
    let western = western.unwrap_or_else(|| DEFAULT_WESTERN_PUBLISHERS.iter().map(|s| s.to_string()).collect());
    (manga, western)
}

/// Manga-detection waterfall (parity with manga-detector.ts, scanner context):
/// manga-publisher list → western-publisher bypass → AniList API cross-reference.
/// The ComicInfo `Manga` tag and the library `isManga` flag are handled by the caller before this.
pub async fn detect_manga(client: &Client, name: &str, publisher: &str, year: i32, manga_pubs: &[String], western_pubs: &[String]) -> bool {
    let pub_lower = publisher.to_lowercase();

    // Step 1: known manga publisher.
    if manga_pubs.iter().any(|mp| pub_lower.contains(mp.as_str())) {
        log::info!("[Manga Engine] Identified via Publisher: {}", publisher);
        return true;
    }

    // Step 2: known western publisher → not manga; skip the AniList call.
    if western_pubs.iter().any(|wp| pub_lower.contains(wp.as_str())) {
        log::info!("[Manga Engine] Bypassing AniList due to Western Publisher: {}", publisher);
        return false;
    }

    if name.is_empty() {
        return false;
    }

    // Step 3: AniList GraphQL cross-reference.
    match check_anilist(client, name, year).await {
        Ok(true) => {
            log::info!("[Manga Engine] Identified via AniList API Match for \"{}\"", name);
            true
        }
        Ok(false) => false,
        Err(e) => {
            log::warn!("[Manga Engine] AniList check failed for \"{}\": {}", name, e);
            false
        }
    }
}

async fn check_anilist(client: &Client, title: &str, release_year: i32) -> anyhow::Result<bool> {
    log::debug!("[Manga Engine Debug] Querying AniList GraphQL API for title: \"{}\"", title);

    let query = r#"query ($search: String) { Page(page: 1, perPage: 3) { media(search: $search, type: MANGA) { title { romaji english } startDate { year } format } } }"#;
    let body = serde_json::json!({ "query": query, "variables": { "search": title } });

    let resp = client
        .post("https://graphql.anilist.co")
        .header("Content-Type", "application/json")
        .header("Accept", "application/json")
        .timeout(Duration::from_secs(10))
        .json(&body)
        .send()
        .await?;

    if !resp.status().is_success() {
        return Ok(false);
    }

    let data: serde_json::Value = resp.json().await?;
    let media = data["data"]["Page"]["media"].as_array().cloned().unwrap_or_default();
    Ok(anilist_media_matches(&media, title, release_year))
}

/// True if any AniList media entry's English/Romaji title equals the search title (case-insensitive),
/// within a ±4-year window of the release year — a same-title match further off is likely a different
/// work. Split out from the HTTP call so the title/year matching can be unit-tested.
fn anilist_media_matches(media: &[serde_json::Value], title: &str, release_year: i32) -> bool {
    let search_title = title.trim().to_lowercase();
    for m in media {
        let eng = m["title"]["english"].as_str().map(|s| s.trim().to_lowercase()).unwrap_or_default();
        let romaji = m["title"]["romaji"].as_str().map(|s| s.trim().to_lowercase()).unwrap_or_default();

        if search_title == eng || search_title == romaji {
            // Fuzzy year guard: a same-title match more than 4 years off is likely a different work.
            if release_year > 0 {
                if let Some(jp_year) = m["startDate"]["year"].as_i64() {
                    if (release_year as i64 - jp_year).abs() > 4 {
                        log::info!("[Manga Engine] AniList match rejected due to year mismatch ({} vs JP {})", release_year, jp_year);
                        continue;
                    }
                }
            }
            return true;
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn anilist_matcher_honors_title_and_year_window() {
        let media = vec![json!({ "title": { "english": "Berserk", "romaji": "Berserk" }, "startDate": { "year": 1989 } })];
        // Exact title, no year constraint → match.
        assert!(anilist_media_matches(&media, "berserk", 0));
        // Within ±4 years → match.
        assert!(anilist_media_matches(&media, "Berserk", 1991));
        // Same title but >4 years off → rejected (likely a different work).
        assert!(!anilist_media_matches(&media, "Berserk", 2010));
        // Different title → no match.
        assert!(!anilist_media_matches(&media, "Naruto", 0));
        // Romaji-only match still counts.
        let romaji_only = vec![json!({ "title": { "english": null, "romaji": "Kingdom" }, "startDate": { "year": 2006 } })];
        assert!(anilist_media_matches(&romaji_only, "kingdom", 2007));
    }

    #[tokio::test]
    async fn detect_manga_short_circuits_on_known_publishers() {
        let client = reqwest::Client::new();
        let manga = vec!["viz media".to_string(), "kodansha".to_string()];
        let western = vec!["marvel".to_string(), "dc comics".to_string()];
        // Known manga publisher → true without hitting AniList.
        assert!(detect_manga(&client, "Any Title", "VIZ Media LLC", 2020, &manga, &western).await);
        // Known western publisher → false (AniList bypassed).
        assert!(!detect_manga(&client, "Any Title", "Marvel Comics", 2020, &manga, &western).await);
    }
}
